src/Common/ZooKeeper/ZooKeeperLoadBalancer.cpp

```diff
@@ -0,0 +1,164 @@
+        host_info.secure = secure;
+        host_info_list.push_back(host_info);
+    }
+}
+
+void ZooKeeperLoadBalancer::setBetterKeeperHostUpdater(BetterKeeperHostUpdater handler)
+{
+    std::lock_guard lock{mutex};
+    better_keeper_host_updater = handler;
+}
+
+
+[[noreturn]] void throwWhenNoHostAvailable(bool dns_error_occurred)
+{
+    if (dns_error_occurred)
+        throw zkutil::KeeperException::fromMessage(Coordination::Error::ZCONNECTIONLOSS, "Cannot resolve any of provided ZooKeeper hosts due to DNS error");
+    else
+        throw zkutil::KeeperException::fromMessage(Coordination::Error::ZCONNECTIONLOSS, "Cannot use any of provided ZooKeeper nodes");
+}
+
+
+void ZooKeeperLoadBalancer::shuffleHosts()
+{
+    std::function<Priority(size_t index)> get_priority = args.get_priority_load_balancing.getPriorityFunc(args.get_priority_load_balancing.load_balancing, 0, args.hosts.size());
+    for (size_t i = 0; i < host_info_list.size(); ++i)
+    {
+        if (get_priority)
+            host_info_list[i].priority = get_priority(i);
+        host_info_list[i].randomize();
+    }
+
+    ::sort(host_info_list.begin(), host_info_list.end(), HostInfo::compare);
+}
+
+void ZooKeeperLoadBalancer::recordKeeperHostError(UInt8 original_index)
+{
+    std::lock_guard lock{mutex};
+    for (auto & host_info : host_info_list)
+    {
+        if (host_info.original_index == original_index)
+        {
+            LOG_DEBUG(log, "Load Balancer records a failure on host {}, original index {}", host_info.address, static_cast<UInt32>(original_index));
+            // TODO: add the actual implementation to lower the priority/scoring of the host.
+            return;
+        }
+    }
+    LOG_ERROR(log, "Does not find host to record the failure with original index {}", static_cast<UInt32>(original_index));
+}
+
+std::unique_ptr<Coordination::ZooKeeper> ZooKeeperLoadBalancer::createClient()
+{
+    std::lock_guard lock{mutex};
+    shuffleHosts();
+    bool dns_error_occurred = false;
+    for (size_t i = 0; i < host_info_list.size(); ++i)
+    {
+        if (!isKeeperHostDNSAvailable(log, host_info_list[i].address, dns_error_occurred))
+            continue;
+
+        LOG_INFO(log, "Connecting to ZooKeeper host {}, number of attempted hosts {}", host_info_list[i].address, i+1);
+        try
+        {
+            auto client  = std::make_unique<Coordination::ZooKeeper>(host_info_list[i].toZooKeeperNode(), args, zk_log);
+            // Non optimal case: we connected to a keeper host that is not at highest priority, set a timeout to force reconnect.
+            // Except for RANDOM and ROUND_ROBIN. For random, as it's random anyway regardless of first or second host.
+            // For round robin, hosts tried later will be in the same order as well.
+            const auto & lb_algorithm = args.get_priority_load_balancing.load_balancing;
+            if (i != 0 && (lb_algorithm != LoadBalancing::RANDOM && lb_algorithm != LoadBalancing::ROUND_ROBIN))
```

Comment:
I do not like that still.
I think that we should think about desire model here and what interface will be good to implement that model.

I want to see the model when lb_algorithm defines which connections are permanent and which are temporary. For example:  HOSTNAME_LEVENSHTEIN_DISTANCE could have several host with equal priority. We do not what do a loop switching between such equal hosts. That decision should not  be done by `createClient` based on just host order.

Reply:
also ROUND_ROBIN,FIRST_OR_RANDOM have mutable priorities , each call `getPriorityFunc` returns different result.

Reply:
Fair point. `HOSTNAME_LEVENSHTEIN_DISTANCE` indeed can have several host with same priority. `i != 0` does not mean we end up with a sub-optimal choice. We need a better interface other than this naive check. Though as discussed before, static algorithms (existing ones) would select hosts purely by algorithm itself; availability zone would require input and interact with ZooKeeperLoadBalancer tracking history. Let me think about that.

Also `getPriorityFunc` is used in clickhouse client connection pool. We probably don't want to change that, for now.

> do we somewhere take getPriorityFunc into account?
Not sure what you meant by that. getPriorityFunc is called outside the loop once per connection, as before.

> also ROUND_ROBIN,FIRST_OR_RANDOM have mutable priorities
ROUND_ROBIN does. But FIRST_OR_RANDOM should be deterministic, we always pass in offset = 0 in the getPriorityFunc call here.
 https://github.com/ClickHouse/ClickHouse/blob/6b35213dd9154dbf773eb1dab0d33e17b8c5df36/src/Common/GetPriorityForLoadBalancing.cpp#L33

Reply:
just get rid off if `getPriorityFunc`.

Reply:
check this @CheSema
[add IClientsConnectionBalancer](https://github.com/ClickHouse/ClickHouse/pull/56563/commits/c189c48b47ed81a13dc9976603db9f802be6457f)

Reply:
@antonio2368 
Since I starting contributing here, I would ask you, please, to review us both. When I open this PR as ready.

Reply:
@incfly please make in draft.